demos/twa-webview-fallback/src/main/java/com/google/browser/examples/twawebviewfallback/WebViewFallbackActivity.java

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.browser.examples.twawebviewfallback;
+
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.browser.customtabs.CustomTabsClient;
+import androidx.browser.customtabs.CustomTabsIntent;
+
+import android.net.Uri;
+import android.os.Bundle;
+import android.view.View;
+import android.view.ViewGroup;
+import android.webkit.RenderProcessGoneDetail;
+import android.webkit.WebResourceRequest;
+import android.webkit.WebSettings;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import android.widget.FrameLayout;
+import android.widget.Toast;
+
+public class WebViewFallbackActivity extends AppCompatActivity {
+    public static final String KEY_LAUNCH_URI =
+            "com.google.browser.examples.twawebviewfallback.WebViewFallbackActivity.KEY_LAUNCH_URL";
+
+    private Uri launchUrl;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        this.launchUrl = this.getIntent().getParcelableExtra(KEY_LAUNCH_URI);
+
+        FrameLayout webViewContainer = new FrameLayout(this);
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(createWebViewClient());
+        webViewContainer.addView(webView,
+                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
+
+        WebSettings webSettings = webView.getSettings();
+        webSettings.setJavaScriptEnabled(true);
+
+        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(
+                FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
+
+        setContentView(webViewContainer,layoutParams);
+        webView.loadUrl(launchUrl.toString());
+    }
+
+    private WebViewClient createWebViewClient() {
```

Comment:
Are we happy with the WebView client having references to the Activity? We probably are, but just want to make sure there's not a memory leak or something.

Reply:
The WebViewClient is attached to the WebView, which is attached to the Activity. When the Activity is GC'd, those will be GC'd with it. Having said that, we can refactor this and pass the launchUrl as a parameter when constructing the WebViewClient.